clientconn.go

```diff
@@ -411,6 +433,9 @@ func (cc *ClientConn) Close() error {
 	}
 	cc.state = Shutdown
 	cc.stateCV.Broadcast()
+	if cc.events != nil {
+		cc.events.Finish()
```

Comment:
What if cc.events.Printf is called after cc.event.Finish() has been called? Is it no-op or there is some issue? If it is an issue, we need to set cc.events to nil here but then we still need a mutex to protect cc.events.

I actually tend to put the operations on cc.events (Printf, errorf, finish ) along with the change of cc.state. cc.state records the state change of ClientConn (and here we just need to push it to tracing) and has been protected by a mutex.